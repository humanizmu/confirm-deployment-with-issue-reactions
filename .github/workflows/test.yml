name: 'build-test'
on: # rebuild any PRs and main branch changes
  push:

jobs:
  build: # make sure build/ci work properly
    timeout-minutes: 2
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - run: |
          npm install
      - run: |
          npm run all
  test: # make sure the action works on a clean machine without building
    timeout-minutes: 2
    runs-on: ubuntu-latest
    # if: github.ref == 'refs/heads/main'
    steps:
      - uses: actions/checkout@v2
      - uses: ./
        # with:
        #   githubToken: ${{ secrets.GITHUB_TOKEN }}
