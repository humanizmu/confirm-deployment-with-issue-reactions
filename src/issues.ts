import * as core from '@actions/core'
import * as github from '@actions/github'
// eslint-disable-next-line import/no-unresolved
import {Api} from '@octokit/plugin-rest-endpoint-methods/dist-types/types'

export type Issues = Api['rest']['issues']

export const getIssueReactions = async (
  issues: Issues,
  issueNumber: number
) => {
  const issue = await issues.get({
    ...github.context.repo,
    issue_number: issueNumber
  })

  return issue.data.reactions
}

export const createIssue = (issues: Issues) => {
  const {GITHUB_SHA, GITHUB_RUN_ID} = process.env
  const commitSha = GITHUB_SHA?.substring(0, 7)

  console.log(github.context)
  console.log('', github.context.repo)

  return issues.create({
    ...github.context.repo,
    title: `🌺 Confirm deployment of ${commitSha}`,
    body: `The confirmation step has been requested by run https://github.com/humanizmu/frontend/actions/runs/${GITHUB_RUN_ID}

Related commit ${commitSha}

To confirm this step react to this issue with 👍
To cancel this step react to the issue with 👎`
  })
}

export const closeIssue = async (
  issues: Issues,
  number: number
): Promise<void> => {
  await issues.update({
    ...github.context.repo,
    issue_number: number,
    state: 'closed'
  })

  core.info(`Closed issue #${number}`)
}
