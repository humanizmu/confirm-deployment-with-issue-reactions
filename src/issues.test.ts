import * as core from '@actions/core'
import * as github from '@actions/github'
import {Issues, closeIssue, createIssue, getIssueReactions} from './issues'
import {castToMock} from './test/type-utils'

jest.mock('@actions/github', () => ({
  context: {
    payload: {
      repository: {
        url: 'http://putin.huilo'
      }
    },
    repo: {
      someRepoData: 'putin huilo'
    }
  }
}))

jest.mock('@actions/core')
const coreInfoMock = castToMock(core.info)

const castToIssues = <T>(issuesMock: T) => issuesMock as unknown as Issues

describe('issues', () => {
  describe('getIssueReactions()', () => {
    const reactions = {['+1']: 1}
    const issue = {data: {reactions}}
    const getIssuesMock = jest.fn().mockResolvedValue(issue)
    const mockIssues = castToIssues({get: getIssuesMock})

    let result: unknown

    beforeEach(async () => {
      result = await getIssueReactions(mockIssues, 2)
    })

    it('makes a correct call to get issues endpoint', () => {
      expect(getIssuesMock).toBeCalledWith({
        issue_number: 2,
        ...github.context.repo
      })
    })

    it('returns issue reactions', () => {
      expect(result).toEqual(reactions)
    })
  })

  describe('createIssue()', () => {
    const issueData = {someIssueProp: 'ok'}
    const createIssueMock = jest.fn().mockResolvedValue(issueData)
    const mockIssues = castToIssues({create: createIssueMock})

    let result: unknown
    let originalEnv: NodeJS.ProcessEnv

    beforeEach(async () => {
      originalEnv = process.env
      process.env = {
        ...originalEnv,
        GITHUB_SHA: '1234567890',
        GITHUB_RUN_ID: 'some-run-id'
      }

      result = await createIssue(mockIssues)
    })

    afterEach(() => {
      process.env = originalEnv
    })

    it('makes a correct call to create issue endpoint', () => {
      expect(createIssueMock.mock.calls[0][0]).toMatchSnapshot()
    })

    it('returns issue number', () => {
      expect(result).toEqual(issueData)
    })
  })

  describe('closeIssue()', () => {
    const updateIssueMock = jest.fn().mockResolvedValue(undefined)
    const mockIssues = castToIssues({update: updateIssueMock})

    describe('always', () => {
      beforeEach(async () => {
        await closeIssue(mockIssues, 2)
      })

      it('makes a correct request to update issue endpoint', () => {
        expect(updateIssueMock).toBeCalledWith({
          ...github.context.repo,
          issue_number: 2,
          state: 'closed'
        })
      })
    })

    describe('when issue update request succeeds', () => {
      beforeEach(async () => {
        await closeIssue(mockIssues, 2)
      })

      it('logs correct message', () => {
        expect(coreInfoMock).toBeCalledWith('Closed issue #2')
      })
    })

    describe('when issue update request fails', () => {
      beforeEach(async () => {
        updateIssueMock.mockRejectedValue(undefined)
        try {
          await closeIssue(mockIssues, 2)
          // eslint-disable-next-line no-empty
        } catch {}
      })

      it('does not log a message', () => {
        expect(coreInfoMock).not.toBeCalled()
      })
    })
  })
})
