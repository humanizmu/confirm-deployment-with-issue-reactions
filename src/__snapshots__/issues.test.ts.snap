// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`issues createIssue() makes a correct call to create issue endpoint 1`] = `
Object {
  "body": "The confirmation step has been requested by run http://putin.huilo/actions/runs/some-run-id

Related commit 1234567

To confirm this step react to this issue with 👍
To cancel this step react to the issue with 👎",
  "someRepoData": "putin huilo",
  "title": "🌺 Confirm deployment of 1234567",
}
`;
